database/mongodb/mongodb.go

```diff
@@ -184,10 +207,70 @@ func (m *Mongo) Drop() error {
 	return m.db.Drop(context.TODO())
 }
 
+func (m *Mongo) ensureLockTable() error {
+	indexes := m.db.Collection(m.config.LockingCollection).Indexes()
+	indexOptions := options.Index().SetUnique(true).SetName("lock_unique_key")
+	indexKeys := bsonx.MDoc{
+		DefaultLockingKey: bsonx.Int32(-1),
+	}
+	_, err := indexes.CreateOne(context.TODO(), mongo.IndexModel{
+		Options: indexOptions,
+		Keys:    indexKeys,
+	})
+	if err != nil {
+		fmt.Println(err)
+		return err
+	}
+	return nil
+}
+
+// ensureVersionTable checks if versions table exists and, if not, creates it.
+// Note that this function locks the database, which deviates from the usual
+// convention of "caller locks" in the MongoDb type.
+func (m *Mongo) ensureVersionTable() (err error) {
+	if err = m.Lock(); err != nil {
+		return err
+	}
+
+	defer func() {
+		if e := m.Unlock(); e != nil {
+			if err == nil {
+				err = e
+			} else {
+				err = multierror.Append(err, e)
+			}
+		}
+	}()
+
+	if err != nil {
+		return err
+	}
+	if _, _, err = m.Version(); err != nil {
+		return err
+	}
+	return nil
+}
+
 func (m *Mongo) Lock() error {
+	newLockObj := lockObj{
+		Key: lockKey,
+	}
+	_, err := m.db.Collection(m.config.LockingCollection).InsertOne(context.TODO(), newLockObj)
```

Comment:
So we looked/talked about this before (and I forget the outcome), maybe we should more carefully look at what happens if a lock can't be acquired, and maybe do something in this method here.  I know that the mysql driver I think waited 10 seconds, and another driver did nothing at all.

The reason I say that is because in our case where we want this on container start up in scaling, I think a fail fast right away is not great and the container start up will have to do some loop and backoff there. 

Something to think about, is whether it would be better if we did _something_ here if the database is locked, instead of pushing that complexity to the container.

Reply:
added exponential backoff